google/ingester.go

```diff
@@ -0,0 +1,327 @@
+type Ingester struct {
+	credentialJSON []byte
+
+	service string
+	project string
+	zone    string
+	region  string
+
+	billing *cloudbilling.APIService
+	compute *compute.Service
+
+	ingestionFilter IngestionFilter
+
+	gcpOptions []option.ClientOption
+
+	err error
+}
+
+// NewIngester will initialize the Ingester for Google
+func NewIngester(ctx context.Context, credentialJSON []byte, service string, project, zone string, opts ...Option) (*Ingester, error) {
+	sid, ok := services[service]
+	if !ok {
+		return nil, ErrNotSupportedService
+	}
+
+	ing := &Ingester{
+		credentialJSON: credentialJSON,
+		service:        sid,
+
+		project: project,
+		zone:    zone,
+		region:  zoneToRegion(zone),
+
+		ingestionFilter: DefaultFilter,
+	}
+
+	for _, opt := range opts {
+		opt(ing)
+	}
+
+	gcpOpts := []option.ClientOption{
+		option.WithCredentialsJSON(credentialJSON),
+	}
+	gcpOpts = append(gcpOpts, ing.gcpOptions...)
+
+	billing, err := cloudbilling.NewService(ctx, gcpOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	compute, err := compute.NewService(ctx, gcpOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	ing.billing = billing
+	ing.compute = compute
+
+	return ing, nil
+}
+
+// Ingest will initialize the process of ingesting and it'll push the price.WithProduct found
+// to the returned channed
+func (ing *Ingester) Ingest(ctx context.Context, chSize int) <-chan *price.WithProduct {
+	results := make(chan *price.WithProduct, chSize)
+
+	go func() {
+		defer close(results)
+
+		// Key is the "machineFamily.(cpu|ram)" ex (e2.cpu) and value the price of it
+		// it is used after the imports of the SKU as we also import the MachineTypes
+		// and to calculate the prices we need this
+		var (
+			machinteTypePrices = make(map[string]price.Price)
+		)
+
+		for sku := range ing.fetchSkus(ctx) {
+			select {
+			case <-ctx.Done():
+				ing.err = ctx.Err()
+				return
+			default:
+			}
+
+			// If the SKU has no price we do not need it
+			// IDK if this is possible but this way we
+			// are sure that we'll not ingest it.
+			// As it's ordered in chronological order (of sku.PricingInfo) we take
+			// the last one as it would be the most current and validate
+			// if it has a PricingExpression
+			if sku.PricingInfo == nil || len(sku.PricingInfo) == 0 || sku.PricingInfo[len(sku.PricingInfo)-1].PricingExpression == nil {
+				continue
+			}
+
+			pi := sku.PricingInfo[len(sku.PricingInfo)-1].PricingExpression
+
+			var family, region, service, group, usage string
+			if sku.Category != nil {
+				family = sku.Category.ResourceFamily
+				service = sku.Category.ServiceDisplayName
+				group = sku.Category.ResourceGroup
+				usage = sku.Category.UsageType
+			}
+			for _, region = range sku.ServiceRegions {
+				// Google has a region named 'global' which means the price is applied fro all regions
+				// to make it works with the setup we have we'll replicate it to the region we are
```

Comment:
The main "issue" is that the queries are generic, not specific for provider, so what you said is something I would really try to avoid, to have a `if provider == google` or something like that.

A solution I thought was to use `Location []string` instead of the current `Location *string` on the `product.Filter` so then as this set on the `google` scope to calculate the  price we could set the `Location: []string{region, "global"}`.

I did not do that  to not break  the implementation and also to have prices generally be 1to1 with region, not global. I also don't know if other providers have it this way or not. And having `Location: []string` makes me think of `AND` not `OR` also makes more sense to get a price for 1 region not multiple regions.

It makes us store more data  onto the DB, that's the only drawback I can see.